c/kastore.c

```diff
@@ -919,8 +893,79 @@ kastore_oput(kastore_t *self, const char *key, size_t key_len, void *array,
             goto out;
         }
     }
+    *ret_item = new_item;
+out:
+    return ret;
+}
+
+static int KAS_WARN_UNUSED
+kastore_bput(kastore_t *self, const char *key, size_t key_len, const void *array,
+    size_t array_len, int type, int flags)
+{
+    int ret = 0;
+    kaitem_t *item;
+    ret = kastore_put_item(self, &item, key, key_len, type, flags);
+    if (ret != 0) {
+        goto out;
+    }
+    item->borrowed_array = array;
+    item->array_len = array_len;
+out:
+    return ret;
+}
+
+int KAS_WARN_UNUSED
+kastore_put(kastore_t *self, const char *key, size_t key_len, const void *array,
+    size_t array_len, int type, int flags)
+{
+    int ret;
+    size_t array_size;
+    void *array_copy = NULL;
+
+    if (type < 0 || type >= KAS_NUM_TYPES) {
+        ret = KAS_ERR_BAD_TYPE;
+        goto out;
+    }
+    if (flags & KAS_BORROWS_ARRAY) {
+        ret = kastore_bput(self, key, key_len, array, array_len, type, flags);
+    } else {
+        array_size = type_size(type) * array_len;
+        array_copy = malloc(array_size == 0 ? 1 : array_size);
+        if (array_copy == NULL) {
+            ret = KAS_ERR_NO_MEMORY;
+            goto out;
+        }
+        memcpy(array_copy, array, array_size);
+        ret = kastore_oput(self, key, key_len, array_copy, array_len, type, flags);
+        if (ret == 0) {
+            /* Kastore has taken ownership of the array, so we don't need to free it */
+            array_copy = NULL;
+        }
+    }
 out:
+    kas_safe_free(array_copy);
+    return ret;
+}
 
+int KAS_WARN_UNUSED
+kastore_oput(kastore_t *self, const char *key, size_t key_len, void *array,
+    size_t array_len, int type, int flags)
+{
+    int ret = 0;
+    kaitem_t *item;
+
+    if (flags != KAS_BORROWS_ARRAY && flags != 0) {
```

Comment:
This doesn't make sense  - KAS_BORROWS_ARRAY is not a valid input to the "own_put" operation. We should error here if flags != 0.

This check should be in kastore_put.